### PR TITLE
Tunnel/v4

### PR DESCRIPTION
**Pull request by victorjulien**
Tunnel handling updates to improve thread safety.

Misc related cleanups.

### CODE REVIEW

**Comment by suricata-qa** [comment]
WARNING:
 
 | field | baseline | test | % |
|--|--|--|--| 
| | SURI_TLPW2_autofp_stats_chk | | |
| .flow.spare | 1961331 | 1859202 | 94.79% | 

Pipeline 16862

**Comment by victorjulien** [comment]
I've been running this on my own nfq+autofp setup for 24hours w/o issues now.

**Comment by suricata-qa** [comment]
WARNING:
 
 | field | baseline | test | % |
|--|--|--|--| 
| | SURI_TLPR1_stats_chk | | |
| .app_layer.tx.pgsql | 55 | 0 | - |
| .app_layer.error.pgsql.parser | 55 | 0 | - | 

Pipeline 16895

File: src/decode.h [hunk 1]
```diff
@@ -1094,6 +1103,26 @@ static inline void DecodeSetNoPacketInspectionFlag(Packet *p)
...
+}
+
+static inline bool PacketIsTunnel(const Packet *p)
+{
+    return (p->ttype != PacketTunnelNone);
+}
+
+static inline bool PacketIsNotTunnel(const Packet *p)
+{
+    return (p->ttype != PacketTunnelNone);
```
Comment (inashivb) [inline src/decode.h:1124]: Is this a typo?

File: src/decode.h [hunk 2]
```diff
@@ -1123,6 +1123,24 @@ static inline bool PacketIsNotTunnel(const Packet *p)
...
+    if (PacketIsTunnelRoot(p) && !PacketTunnelIsVerdicted(p) && !outstanding) {
+        SCLogDebug("root %p: verdict", p);
+        return true;
+    } else if (PacketIsTunnelChild(p) && outstanding == 1 && p->root &&
+               PacketTunnelIsVerdicted(p->root)) {
+        SCLogDebug("tunnel %p: verdict", p);
+    } else {
+        return false;
+    }
+    return true;
```
Comment (inashivb) [inline src/decode.h:1142]: nit: could it be
```
        return true;
    }
return false;
```
?

File: src/decode.h (hunk 1 above)
Comment (victorjulien) [reply src/decode.h:1124]: Looks like it, yes.

File: src/decode.h [hunk 3]
```diff
@@ -1091,6 +1104,44 @@ static inline void DecodeSetNoPacketInspectionFlag(Packet *p)
...
+
+static inline bool VerdictTunnelPacketInternal(const Packet *p)
+{
+    const uint16_t outstanding = TUNNEL_PKT_TPR(p) - TUNNEL_PKT_RTV(p);
+    SCLogDebug("tunnel: outstanding %u", outstanding);
+
+    /* if there are packets outstanding, we won't verdict this one */
+    if (PacketIsTunnelRoot(p) && !PacketTunnelIsVerdicted(p) && !outstanding) {
+        SCLogDebug("root %p: verdict", p);
+        return true;
```
Comment (jasonish) [inline src/decode.h:1135]: Remove this return.

**Review by jasonish — CHANGES_REQUESTED** [review]
Some unresolved comments in here. Also seems large enough for a ticket?

**Comment by jasonish** [comment]
> I've been running this on my own nfq+autofp setup for 24hours w/o issues now.

Any thoughts on getting more soak testing?

**Comment by victorjulien** [comment]
> > I've been running this on my own nfq+autofp setup for 24hours w/o issues now.
> 
> Any thoughts on getting more soak testing?

NFQ has seen the most changes, so would be the best one to test. Then we need to make sure the tunnel logic gets triggered a lot. Not sure what the easiest way to set that up would be. Perhaps something like
```
[ client ] < gre tunnel > [ DuT ] < gre tunnel > [server ]
```
Then run something like `ab`.

File: src/decode.h (hunk 2 above)
Comment (jufajardini) [reply src/decode.h:1142]: Wouldn't Jason's suggestion and yours be conflicting, here?

Comment (inashivb) [reply src/decode.h:1142]: I think they'd both mean the same thing.

Comment (jufajardini) [reply src/decode.h:1142]: Oh! I suck at this xD

**Comment by victorjulien** [comment]
replaced by #10597

**Pull request closed without merging**